### PR TITLE
multiplexer: Separate the mux and improve the variant-ids [v0.3]

### PR DESCRIPTION
**Pull request by ldoktor**
This proof-of-concept PR addresses the discussion from https://github.com/avocado-framework/avocado/pull/1535 and extends that PR. The main goal is to completely separate the yaml_to_mux parser, create multiplexer API without hardcoded dependency on mux and then improve the variant IDs.

It's still not intended for inclusion, because there is one big question. Do we want the variants-api to get the default params and incorporate them into the variant, or do we want the variant-api to get the variant and then inject the default params to it? This PR uses the first approach, which is way faster and simpler as it allows the variant plugin to first inject the default params and then to generate the variants already containing the default params. I explored the later version as well, but it was not really nice as it requires the variant-api to get the variant and merge the default_params to the variant per-each variant (which means deep-copy of the variant to avoid altering of the other variants).

Another question is whether I should remove the `TreeNode.ctrl`, which probably shouldn't be used by the default_params.

Last question is regarding the new version, would it be OK to update the documentation not per each change, but rather on one or two stages, because I basically need to re-create the documentation as we separate most of the things described there.

__An important message is that `replay` and `diff` of older jobs is not supported.__

v0: https://github.com/avocado-framework/avocado/pull/1535
v0.2: https://github.com/avocado-framework/avocado/pull/1553

Changes:

```
v0.2: Everything except of the concept of variant-ids (which uses order-by-path instead
      of order by position in yaml file and the patches are split.
v0.3: Rebased
v0.3: Fixed all bugs in order to pass selftests (and adjust some of the selftests)
v0.3: Added a commit to rename multiplexer to variants
```

### CODE REVIEW

**Comment by adereis** [comment]
On Wed, Nov 02, 2016 at 10:35:36AM -0700, Lukáš Doktor wrote:
> This proof-of-concept PR addresses the discussion from
> https://github.com/avocado-framework/avocado/pull/1535 and
> extends that PR. The main goal is to completely separate the
> yaml_to_mux parser, create multiplexer API without hardcoded
> dependency on mux and then improve the variant IDs.
> 
> It's still not intended for inclusion, because there is one big
> question. Do we want the variants-api to get the default params
> and incorporate them into the variant, or do we want the
> variant-api to get the variant and then inject the default
> params to it? This PR uses the first approach, which is way
> faster and simpler as it allows the variant plugin to first
> inject the default params and then to generate the variants
> already containing the default params. I explored the later
> version as well, but it was not really nice as it requires the
> variant-api to get the variant and merge the default_params to
> the variant per-each variant (which means deep-copy of the
> variant to avoid altering of the other variants).

Let me write what we discussed on the call, to clarify what I
propose:

Option 1:
  The "Variants API" (VAPI) receives default parameters from
  Avocado, provides them to plugins and receive back a list of
  Variants, each identified by a Variant-ID (VID). This list is
  provided "AS IS" back to Avocado.

  IMPORTANT: The VAPI doesn't know anything about how VIDs are
  generated, but in this option it's mandatory that the variant
  plugins merge the default params in the resulting Variants.

  Implementation detail of the multiplexer plugin:

  1.1 The VID generated by the multiplexer contains the
  names of variables (leaves) from the default parameters. I
  don't like this option, because this will result in a polluted
  namespace that will be the same for all variants per run.

  1.2 The multiplexer completely ignores the default variables
  when creating the VID.

  1.3 The multiplexer doesn't add the variable names to the VID,
  but it uses their values in the hash computation (which is part
  of the VID).

Option 2:
  VAPI receives from the plugins a list of Variants, each
  identified by a VID. This list is provided "AS IS" back to
  Avocado. When tests request parameters, Avocado will (somehow,
  this is an implementation detail) provide both parameters from
  the Variant and the defaults to tests.

My comments:

  * I would NACK v1.1 from a usability perspective.
  * For the end user (or test-writer) 1.2 and 2 are the same and
    as described above, both match a good architecture, although
    the API of 2 is much cleaner.
  * With 1.3 (and 1.1), Variant-IDs will vary based on what's
    configured (plugins, configuration options, etc), making it
    hard to reuse the Variant-IDs between different
    executions (in test data dirs or command-line options, for
    example). This will also decouple the VID from the
    multiplexer, making it more confuse to the user, who will
    have to understand why the VIDs are changing if they haven't
    changed anything in the main mechanism used for Variant
    generation (the multiplexer).

So I would stick to 2 or 1.2 (doesn't really matter, as long as
the code follows the architectural guidelines above). For debug
purposes, I think a hash of the default parameters (or even the
full dictionary) should be included in the logs somewhere, but
they should not be in the Variant-ID.

> 
> Another question is whether I should remove the
> `TreeNode.ctrl`, which probably shouldn't be used by the
> default_params.

I'm not familiar with the code, so I don't know. But the check is
relatively simple: draw the boxes for each component (the API
itself is a box) and make sure they are properly isolated based
on dependencies. Meaning: a module that doesn't depend on another
doesn't know or care about what happens there.  And of course,
there should be no circular dependencies.

> 
> Last question is regarding the new version, would it be OK to
> update the documentation not per each change, but rather on one
> or two stages, because I basically need to re-create the
> documentation as we separate most of the things described
> there.

I would have to review the details to make sure.  On a first
sight, I think you could remove the obsolete parts of the
documentation in each commit, but add the complete version at the
end.

If this is too much trouble, fixing the documentation at the end
is probably OK IMO.

> 
> __An important message is that `replay` and `diff` of older
> jobs is not supported.__

That's OK.


-- 
Ademar Reis
Red Hat

^[:wq!

**Comment by ldoktor** [comment]
Dne 3.11.2016 v 14:10 Ademar de Souza Reis Jr napsal(a):
> On Wed, Nov 02, 2016 at 10:35:36AM -0700, Lukáš Doktor wrote:
>> This proof-of-concept PR addresses the discussion from
>> https://github.com/avocado-framework/avocado/pull/1535 and
>> extends that PR. The main goal is to completely separate the
>> yaml_to_mux parser, create multiplexer API without hardcoded
>> dependency on mux and then improve the variant IDs.
>>
>> It's still not intended for inclusion, because there is one big
>> question. Do we want the variants-api to get the default params
>> and incorporate them into the variant, or do we want the
>> variant-api to get the variant and then inject the default
>> params to it? This PR uses the first approach, which is way
>> faster and simpler as it allows the variant plugin to first
>> inject the default params and then to generate the variants
>> already containing the default params. I explored the later
>> version as well, but it was not really nice as it requires the
>> variant-api to get the variant and merge the default_params to
>> the variant per-each variant (which means deep-copy of the
>> variant to avoid altering of the other variants).
>
> Let me write what we discussed on the call, to clarify what I
> propose:
>
> Option 1:
> The "Variants API" (VAPI) receives default parameters from
> Avocado, provides them to plugins and receive back a list of
> Variants, each identified by a Variant-ID (VID). This list is
> provided "AS IS" back to Avocado.
>
> IMPORTANT: The VAPI doesn't know anything about how VIDs are
> generated, but in this option it's mandatory that the variant
> plugins merge the default params in the resulting Variants.
>
> Implementation detail of the multiplexer plugin:
>
> 1.1 The VID generated by the multiplexer contains the
> names of variables (leaves) from the default parameters. I
> don't like this option, because this will result in a polluted
> namespace that will be the same for all variants per run.
>
> 1.2 The multiplexer completely ignores the default variables
> when creating the VID.
This is the  current implementation

>
> 1.3 The multiplexer doesn't add the variable names to the VID,
> but it uses their values in the hash computation (which is part
> of the VID).
>
> Option 2:
> VAPI receives from the plugins a list of Variants, each
> identified by a VID. This list is provided "AS IS" back to
> Avocado. When tests request parameters, Avocado will (somehow,
> this is an implementation detail) provide both parameters from
> the Variant and the defaults to tests.
I have this implementation in my branch, but it is really ugly as it 
requires lots of merging and copying of variants. There are shortcuts 
which can be made, but in the end it'll always be harder task, because 
you can't merge the branches before multiplexation.

>
> My comments:
>
> * I would NACK v1.1 from a usability perspective.
> * For the end user (or test-writer) 1.2 and 2 are the same and
> as described above, both match a good architecture, although
> the API of 2 is much cleaner.
> * With 1.3 (and 1.1), Variant-IDs will vary based on what's
> configured (plugins, configuration options, etc), making it
> hard to reuse the Variant-IDs between different
> executions (in test data dirs or command-line options, for
> example). This will also decouple the VID from the
> multiplexer, making it more confuse to the user, who will
> have to understand why the VIDs are changing if they haven't
> changed anything in the main mechanism used for Variant
> generation (the multiplexer).
>
> So I would stick to 2 or 1.2 (doesn't really matter, as long as
> the code follows the architectural guidelines above). For debug
> purposes, I think a hash of the default parameters (or even the
> full dictionary) should be included in the logs somewhere, but
> they should not be in the Variant-ID.
>
>>
>> Another question is whether I should remove the
>> `TreeNode.ctrl`, which probably shouldn't be used by the
>> default_params.
>
> I'm not familiar with the code, so I don't know. But the check is
> relatively simple: draw the boxes for each component (the API
> itself is a box) and make sure they are properly isolated based
> on dependencies. Meaning: a module that doesn't depend on another
> doesn't know or care about what happens there. And of course,
> there should be no circular dependencies.
>
The `avocado.core.tree` is a basic implementation of tree structure used 
to store the params. It has several features like storing key/value 
pairs, providing environment which compounds of key/value pairs 
inherited from antecedent nodes occasionally merging list values. One of 
the features is `TreeNode.ctrl` which allows storing control tags, which 
can influence the node on various occasions. The basic implementation 
allows `remove_node` and `remove_value`, they both act during merging 
the trees when it removes the value/node of the original node.

The motivation was to allow completely remove node on merge, here it 
would allow one to remove the default_parameter (node or value). The 
question is whether we find this feature as core params API or not. I 
think it should be removed and not allowed.

>>
>> Last question is regarding the new version, would it be OK to
>> update the documentation not per each change, but rather on one
>> or two stages, because I basically need to re-create the
>> documentation as we separate most of the things described
>> there.
>
> I would have to review the details to make sure. On a first
> sight, I think you could remove the obsolete parts of the
> documentation in each commit, but add the complete version at the
> end.
>
> If this is too much trouble, fixing the documentation at the end
> is probably OK IMO.
>
Basically right now we describe multiplexation mainly in `Mux` section. 
After this PR I have to split this and create `Variants` section to 
describe the variants, default params and the avocado params system. 
Then I'll have to create `yaml_to_mux plugin` section to describe the 
whole process of multiplexation.

>>
>> __An important message is that `replay` and `diff` of older
>> jobs is not supported.__
>
> That's OK.
>
>
Thank you for nice overview,
Lukáš

**Comment by ldoktor** [comment]
After discussion with guys, let me post here the options:

1.2 - the current implementation, works flawlessly, but it requires passing the default params
2 - works well in normal schema, requires deep magic for debug mode (due to multiple merges it appends the debug information twice). Another problem is that, unless the 1.2, it does not allow optimization before yielding the variants as the plugin does not know the defaults. Due to this limitation it always needs do the full merging per each variant (per each test)
2b - is lightweight version of (2). It only merges default params leaves into variant leaves. This is way faster as we don't have to walk the tree, but on some occasions it moves the value to different location (still the new location should be expected and should not matter). I'll provide example bellow.

In terms of speed the (1.2) is the fastest, because it allows merging the default params before iterating the tests (therefor it does it once per all variants per all tests. The (2b) is significantly slower as it requires updating the default params per each variant and per each test, but it only updates the leaves. The (2) is the slowest, because it does the full tree-merge per each variant and test.

To speed up things we could remember the variants, so they are only calculated once per each variant and repeat them per all tests, but we can't optimize the per-variant merging (basically the difference between 1 and 2 is that by design 1 allows `(1 + 2 + 3) * 10`, while the 2 allows only `1 * 10 + 2 * 10 + 3 * 10`.

Now the promised example of 2b:

```yaml
# Default params
virt:
    bin: /bin/qemu-kvm
foo:
    bar:baz


# my.yaml
virt:
    bin: !mux
        default:
        upstream:
            bin: /bin/qemu-system-x86_64


# The 1.2 and 2 produces this (--dry-run)
/virt:bin:/bin/qemu-kvm
/virt/bin/upstream:bin:/bin/qemu-system-x86_64
/foo:bar:baz

# The 2b produces this (--dry-run)
/virt/bin/default:bin:/bin/qemu-kvm
/virt/bin/upstream:bin:/bin/qemu-system-x86_64
/foo:bar:baz
```
It shouldn't be a problem, because the leaf path is after the expected location and one would be unable to query for that particular location as AvocadoParams only returns environments of leaves. So the correct query for `bin` would be:

```python
params.get("/virt/*", "bin")
params.get("*", "bin")
```
and it works well in all cases as the `/virt/bin/default` matches that.

Anyway, to wrap it up, the (2) is the cleanest way as it does not depend on the variant plugin to merge the defaults (which can be source of issues), the (2b) also avoids that, but could be a bit confusing and the (1.2) has the biggest optimization potential, but a broken variant plugin could ignore/corrupt default params.

This PR implements (1.2) and I think it's the best one as it allows the greatest optimization (in the future).

File: avocado/plugins/yaml_to_mux.py [hunk 1]
```diff
@@ -307,12 +382,31 @@ def _log_deprecation_msg(deprecated, current):
...
+            if mux_filter_only:
+                args.mux_filter_only = mux_filter_only + only
+            else:
+                args.mux_filter_only = only
+        out = getattr(args, "filter_out", None)
+        if out:
+            self._log_deprecation_msg("--filter-out", "--mux-out")
+            mux_filter_out = getattr(args, "mux_filter_out")
+            if mux_filter_out:
+                args.mux_filter_only = mux_filter_out + out
```
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: `s/args.mux_filter_only/args.mux_filter_out`

File: avocado/core/multiplexer.py [hunk 2]
```diff
@@ -397,11 +397,8 @@ def parse(self, args):
```
Comment (clebergnu) [inline avocado/core/multiplexer.py]: The wording on the commit message is a bit confusing.  It says: "only the multiplexer plugin should produce variants".  This does *not* match the archicture we settled upon.  It can be seen on the following diagram:

https://github.com/avocado-framework/avocado/pull/1535#issuecomment-253927016

That "Other variant plugin(s)" can provide variants.  Can you better explain the commit message?

Also, WRT to the example given with Avocado-Virt, it would be nice to use the internal component names .  For instance: "this version does the filtering only on the multiplexed tree (`MuxTree`) ... ".  Avocado-Virt can be mentioned as one user of that interface, or even skipped altogether.

File: avocado/core/tree.py [hunk 3]
```diff
@@ -104,6 +104,10 @@ def __eq__(self, other):
                     return False
             return True
 
+    def __ne__(self, other):
+        """ Inverted eq """
+        return not self == other
```
Comment (ldoktor) [inline avocado/core/tree.py]: I don't think so, or at least it does not work properly in all cases. I added this commit because it didn't work later in the code...

File: avocado/core/tree.py [hunk 4]
```diff
@@ -71,15 +71,14 @@ def __init__(self, name='', value=None, parent=None, children=None):
...
         self._environment = None
         self.environment_origin = {}
-        self.ctrl = []
-        self.multiplex = None
         for child in children:
             self.add_child(child)
 
     def __repr__(self):
-        return 'TreeNode(name=%r)' % self.name
+        return 'SimpleTreeNode(name=%r)' % self.name
```
Comment (ldoktor) [inline avocado/core/tree.py]: Sorry, this is leftover from another version.

File: avocado/core/tree.py [hunk 5]
```diff
@@ -510,7 +505,7 @@ def process_node(node):
         Generate this node's tree-view
         :return: list of lines
         """
-        if node.multiplex:
+        if getattr(node, "multiplex", None):
```
Comment (ldoktor) [inline avocado/core/tree.py]: I'd prefer keeping it this way for now, because there are already so many changes. When we have this part working we can address the tree-view (as I mentioned somewhere I'd prefer having `tree_view` method inside the tree-class, but let's explore the possibilities after those changes).

File: avocado/core/tree.py [hunk 6]
```diff
@@ -469,9 +464,9 @@ def merge(self, other):
         return super(TreeNodeDebug, self).merge(other)
 
 
-def get_named_tree_cls(path):
+def get_named_tree_cls(path, klass=TreeNodeDebug):
```
Comment (ldoktor) [inline avocado/core/tree.py]: OK, I'll try to extract it in the next v.

File: avocado/plugins/yaml_to_mux.py [hunk 7]
```diff
@@ -59,7 +59,51 @@ class ListOfNodeObjects(list):     # Few methods pylint: disable=R0903
...
+    Class for bounding nodes into tree-structure with support for
+    multiplexation
+    """
+
+    def __init__(self, name='', value=None, parent=None, children=None):
+        super(MuxTreeNode, self).__init__(name, value, parent, children)
+        self.multiplex = None
+
+    def __repr__(self):
+        return 'TreeNode(name=%r)' % self.name
```
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: Yep, it's a leftover of the original. I'll fix it before moving.

File: avocado/plugins/yaml_to_mux.py [hunk 8]
```diff
@@ -59,7 +59,51 @@ class ListOfNodeObjects(list):     # Few methods pylint: disable=R0903
...
+    def merge(self, other):
+        """
+        Merges `other` node into this one without checking the name of the
+        other node. New values are appended, existing values overwritten
+        and unaffected ones are kept. Then all other node children are
+        added as children (recursively they get either appended at the end
+        or merged into existing node in the previous position.
+        """
+        super(MuxTreeNode, self).merge(other)
+        if other.multiplex is True:
```
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: This is quite tricky question. The multiplex is an important tag changing the way the tree is processed. When merging two trees, we might end up in a situation where simpler tree is merged into a more-specific tree, where the simpler tree adds some default values, but the more-specific tree already contains it's variants. Then we don't want to loose the variants:

```
# simpler tree
cpus:
    count: 4
    type: x86_64

# complex tree
cpus: !mux
    x86_64:
        type: x86_64
    ppc64:
        type: ppc64

# should produce
[medic@w541 avocado ]$ avocado multiplex -m simple.yaml complex.yaml -c
Variants generated:

Variant 1:    /run/cpus/x86_64
    /run/cpus/x86_64:type => x86_64
    /run/cpus:count       => 4

Variant 2:    /run/cpus/ppc64
    /run/cpus/ppc64:type => ppc64
    /run/cpus:count      => 4
[medic@w541 avocado ]$ avocado multiplex -m complex.yaml simple.yaml -c
Variants generated:

Variant 1:    /run/cpus/x86_64
    /run/cpus/x86_64:type => x86_64
    /run/cpus:count       => 4

Variant 2:    /run/cpus/ppc64
    /run/cpus/ppc64:type => ppc64
    /run/cpus:count      => 4

# but with `self.multiplex = other.multiplex` only the value of the later is kept
[medic@w541 avocado ]$ avocado multiplex -m complex.yaml simple.yaml -c
Variants generated:

Variant 1:    /run/cpus/x86_64, /run/cpus/ppc64
    /run/cpus/ppc64:type  => ppc64
    /run/cpus/x86_64:type => x86_64
    /run/cpus:count       => 4
```

File: avocado/plugins/yaml_to_mux.py [hunk 9]
```diff
@@ -59,7 +59,51 @@ class ListOfNodeObjects(list):     # Few methods pylint: disable=R0903
...
+    Debug version of TreeNodeDebug
+    :warning: Origin of the value is appended to all values thus it's not
+    suitable for running tests.
+    """
+
+    def __init__(self, name='', value=None, parent=None, children=None,
+                 srcyaml=None):
+        MuxTreeNode.__init__(self, name, value, parent, children)
+        tree.TreeNodeDebug.__init__(self, name, value, parent, children,
+                                    srcyaml)
```
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: Yes, I know I have to rename it to source or origin, but I'd prefer first having everything all set and then focus one style fixes.

File: avocado/plugins/yaml_to_mux.py [hunk 10]
```diff
@@ -293,10 +350,28 @@ def configure(self, parser):
             mux.add_argument("-m", "--mux-yaml", nargs='*', metavar="FILE",
                              help="Location of one or more Avocado"
                              " multiplex (.yaml) FILE(s) (order dependent)")
+            mux.add_argument('--mux-filter-only', nargs='*', default=[],
```
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: PS: If you agree with this approach, I can create cards for this, the style fixes and some other non-critical improvements.

File: avocado/core/multiplexer.py (hunk 2 above)
Comment (ldoktor) [inline avocado/core/multiplexer.py]: This commit is a prep work to get to the point where we can follow the diagram. What the commit message tries to say is that with this change the `core` API should have no knowledge about producing the variants (only the `Mux` allows to get variants from the `yaml_to_mux`. I'll try to improve the message, though... The same with avocado-virt. I used commit messages to store my ideas, before the final release I have to adjust them all.

File: avocado/plugins/yaml_to_mux.py [hunk 11]
```diff
@@ -392,17 +393,27 @@ def __init__(self, root, mux_path):
...
+            variant.sort(key=lambda x: x.path)
+            fingerprint = "-".join(_.fingerprint() for _ in variant)
+            variant_ids.append("-".join(node.name for node in variant) + '-' +
+                               sha.sha(fingerprint).hexdigest()[:4])
+        return variant_ids
 
     def __iter__(self):
-        for i, variant in enumerate(self.variants, 1):
+        for vid, variant in itertools.izip(self.variant_ids, self.variants):
             '''
```
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: Looks like I missed the triple quotes here from previous commit.  This is related to the discussion on where to apply the default parameters, right? I mean:

https://github.com/avocado-framework/avocado/pull/1591#issuecomment-258697795

I feel that the current implementation (1.2) is adequate.  The "requirement" on default_parameters is OK, as long as it can be empty, and I saw no reason it can't.

File: avocado/core/runner.py [hunk 12]
```diff
@@ -501,7 +501,7 @@ def run_suite(self, test_suite, mux, timeout=0, replay_map=None):
         else:
             deadline = None
 
-        test_result_total = mux.get_number_of_tests(test_suite)
+        test_result_total = len(mux) * len(test_suite)
```
Comment (clebergnu) [inline avocado/core/runner.py]: OK.

File: avocado/plugins/yaml_to_mux.py [hunk 13]
```diff
@@ -417,6 +417,12 @@ def __iter__(self):
                    "variant": variant,
                    "mux_path": self.mux_path}
 
+    def __len__(self):
+        """
+        Report the number of variants
+        """
+        return len(self.variant_ids)
```
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: OK, since `self.variant_ids` is set at `__init__` there should be no issues here.

File: selftests/unit/test_multiplexer.py [hunk 14]
```diff
@@ -39,7 +39,7 @@ def test_basic(self):
         self.assertNotEqual(self.params1, self.params2)
         repr(self.params1)
         str(self.params1)
-        str(multiplexer.AvocadoParams([], 'Unittest', [], {}))
+        str(variants.AvocadoParams([], 'Unittest', [], {}))
```
Comment (clebergnu) [inline selftests/unit/test_multiplexer.py]: OK, although I still find a few references to the term "multiplexer".  Hopefully this would be helpful for you to check if they make sense or not:

```
$ git grep multiplexer 
README.rst:- A multiplexer that allows tests to be run with different sets of variables;
avocado/core/runner.py:        :param mux: the multiplexer.
avocado/plugins/diff.py:            return ["Unsupported version of multiplexer: %s" % details]
docs/source/Configuration.rst:* multiplexer
docs/source/Mux.rst:The ``Mux`` object is defined in ``avocado/core/multiplexer.py``, is always
docs/source/Mux.rst:by path. But the most powerful multiplexer feature is that it can generate
docs/source/Mux.rst:time.In order to generate all possible variants multiplexer creates cartesian
docs/source/ReferenceGuide.rst:The multiplexer component creates different sets of variables
docs/source/ReferenceGuide.rst:multiplexer to identify each variant. It should be unique per
docs/source/ReferenceGuide.rst:variant inside a set. In other words, the multiplexer generates a
docs/source/ReferenceGuide.rst:A simpler implementation of the multiplexer uses serial integers
docs/source/ReferenceGuide.rst:.. note:: The multiplexer supports serialized variant IDs only
docs/source/WritingTests.rst:   2016-11-02 11:13:01,394 multiplexer      L0166 DEBUG| PARAMS (key=timeout, path=*, default=None) => 3
docs/source/WritingTests.rst:   2016-11-02 11:13:01,396 multiplexer      L0166 DEBUG| PARAMS (key=sleep_length, path=*, default=1) => 5
docs/source/WritingTests.rst:If you pass that multiplex file to the runner multiplexer, this will register
```

File: avocado/core/job.py [hunk 15]
```diff
@@ -377,6 +377,8 @@ def _log_avocado_datadir():
         job_log.info('')
```
Comment (ldoktor) [inline avocado/core/job.py]: For now it won't be used, but eventually the avocado-virt should use it. I'll add selftests for now...

File: avocado/plugins/yaml_to_mux.py [hunk 16]
```diff
@@ -432,7 +538,11 @@ def run(self, args):
...
 
+        mux_path = getattr(args, 'mux_path', None)
+        if mux_path is None:
+            mux_path = ['/run/*']
+
         data = apply_filters(data, getattr(args, 'mux_filter_only', None),
                              getattr(args, 'mux_filter_out', None))
         if data != MuxTreeNode():
-            args.mux.data_merge(data)
+            args.mux.add_variants_plugin(MuxPlugin(data, mux_path))
```
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: Yes, this is just a prep-work. I think it's more important to get this working as it improves the handling and then continue improving it. First we need to know, whether it's possible to separate it and when we have somehow stable API we should focus on making it properly modular. Defining single `add_*_plugin` should be enough to then turn it into a proper plugin and I think it's fine for now.

File: avocado/core/tree.py [hunk 17]
```diff
@@ -107,6 +107,12 @@ def __ne__(self, other):
         """ Inverted eq """
         return not self == other
 
+    def fingerprint(self):
+        """
+        Reports string which represents the value of this node.
+        """
+        return str(self.path) + str(self.environment) + str(self.ctrl)
```
Comment (ldoktor) [inline avocado/core/tree.py]: sure, btw the speed of formatting vs. adding is not 100% like that, it depends on many factors, but I can use it that way. And yes, the `%s` means convert it to string, the conversion is only necessary when using the `+`.

File: avocado/plugins/yaml_to_mux.py [hunk 18]
```diff
@@ -18,6 +18,7 @@
 import logging
 import os
 import re
+import sha
```
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: The ordering notice is quite important. What it means is, that:

```yaml
first:
    foo: bar
second:
    bar: baz
```

```yaml
second:
    bar: baz
first:
    foo: bar
```

Produces the same variant-id. Without the manual sorting it'd produce different ids because the multiplexer is ordered (therefor the order of `first` and `second` matters inside the variant, even though the final params are the same)

The default_params has no relation to the variant-id whatsoever.

File: avocado/plugins/yaml_to_mux.py (hunk 11 above)
Comment (ldoktor) [inline avocado/plugins/yaml_to_mux.py]: yep, this is a remnant to not forget the version (2) is also possible and will not be part of the final version.

File: selftests/unit/test_multiplexer.py (hunk 14 above)
Comment (ldoktor) [inline selftests/unit/test_multiplexer.py]: Sure, this commit is marked as FIXME, it's only meant as RFC whether the name Variants would be OK and demonstration how would this change look like in the code.

File: avocado/core/tree.py (hunk 3 above)
Comment (clebergnu) [inline avocado/core/tree.py]: It does.  I added a breakpoint to `__ne__` and it gets called from those asserts.

File: avocado/plugins/yaml_to_mux.py [hunk 19]
```diff
@@ -59,7 +59,51 @@ class ListOfNodeObjects(list):     # Few methods pylint: disable=R0903
     pass
 
 
-def _create_from_yaml(path, cls_node=tree.TreeNode):
+class MuxTreeNode(tree.TreeNode):
```
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: OK.

File: avocado/plugins/yaml_to_mux.py (hunk 8 above)
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: Interesting.  Thanks for the explanation!

File: avocado/plugins/yaml_to_mux.py (hunk 10 above)
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: I think it's OK to have filtering at the top most level (at the variants API) and cascading those to the individual implementations.  It's actually nice if we can have a neat model that accommodates all of them uniformly.

File: avocado/core/multiplexer.py (hunk 2 above)
Comment (clebergnu) [inline avocado/core/multiplexer.py]: Sure, it's fair to use commit message as a scratch pad.  As we move along the review process, it is also a good idea to think of other people reading the message and correlating that to the code.  It's a good idea to evolve the message.

File: avocado/core/job.py [hunk 20]
```diff
@@ -376,34 +375,16 @@ def _log_avocado_datadir():
         job_log.info('logs     ' + data_dir.get_logs_dir())
         job_log.info('')
 
-    def _log_mux_tree(self, mux):
```
Comment (clebergnu) [inline avocado/core/job.py]: Again, I may be missing something, but I had the idea that a tree would always be present and be the source of parameters.  Looking at the diagram:

("->" means "depends on" or "knows about")

```
   +------------------------------+
   | Test                         |
   +------------------------------+
             |
             |
   +---------v---------+    +--------------------------------+
   | Parameters System |--->| Variants Generation Plugin API |
   +-------------------+    +--------------------------------+
             ^                                ^
             |                                |
   +--------------------------------------+   |
   | +--------------+ +-----------------+ |   |
   | | avocado-virt | | other providers | |   |
   | +--------------+ +-----------------+ |   |
   +--------------------------------------+   |
                                              |
                 +----------------------------+-----+
                 |                                  |
                 |                                  |
                 |                                  |
       +--------------------+           +-------------------------+
       | Multiplexer Plugin |           | Other variant plugin(s) |
       +--------------------+           +-------------------------+
             |
             |
       +-----v---------------------------+
       | +------------+ +--------------+ |
       | | --mux-yaml | | --mux-inject | |
       | +------------+ +--------------+ |
       +---------------------------------+
```

Isn't the parameter system backed by a Tree? The only other option I can see, which looks incomplete, is if `default_params` would be (but is not) a replacement for it.

File: avocado/core/multiplexer.py [hunk 21]
```diff
@@ -20,66 +20,12 @@
...
-    Object representing part of the tree from the root to leaves or another
-    multiplex domain. Recursively it creates multiplexed variants of the full
-    tree.
-    """
-
-    def __init__(self, root):
-        """
-        :param root: Root of this tree slice
-        """
-        self.root = root
```
Comment (clebergnu) [inline avocado/core/multiplexer.py]: This is pretty minor, and it's also something core, so users wouldn't be affect.  It just serves as a best practice reminder.

File: avocado/core/tree.py (hunk 17 above)
Comment (clebergnu) [inline avocado/core/tree.py]: OK.

File: avocado/plugins/yaml_to_mux.py (hunk 18 above)
Comment (clebergnu) [inline avocado/plugins/yaml_to_mux.py]: OK, understood.

**Comment by clebergnu** [comment]
@ldoktor do you need any more input from me on this version?  I'm not sure, but I had the impression that you posted some other comments (I think I saw them in a mobile app), but then I don't see them here.

**Comment by clebergnu** [comment]
This has many conflicts with master, and the mux-rename3 has been merged, so I'm closing this one also.

**Comment by ldoktor** [comment]
Well this depends on how do we want to pull requests use. I prefer having the latest versions still opened while someone is working on a next versions, therefor the information that this function is in progress and it's been working on now. By this terminology I'd keep this open as this PR demonstrates the idea of variant-ids, which hasn't been implemented yet. There were several stages already merged, but they did not reflect this feature and there might be people being interested in this particular feature and they still can review it here or just add notes regarding this feature. If we close it, than no one knows this feature is being worked on and they can't provide feedback/ideas until I finish a new version, which is IMO less productive.

Take avocado-vt for example, the PRs are opened until they are merged, or completely rejected. We use new PRs for new versions, so the question is whether we do want to close the old ones when new PR is opened, or when it becomes a bit outdated (for whatever reason).

@apahim @clebergnu what do you guys think?

**Comment by ldoktor** [comment]
[[mux] Improve the generation of Variant-IDs](https://trello.com/c/1lzDCiqn/835-mux-improve-the-generation-of-variant-ids)
